scripts/coverage.sh

```diff
@@ -41,15 +57,60 @@ push_to_coveralls() {
     goveralls -coverprofile="$profile"
 }
 
-generate_cover_data $(go list ./... | grep -v /vendor/)
+while [[ $# -gt 0 ]]; do
+    case "$1" in
+        --html)
+            html=true
+            shift
+            ;;
+        --coveralls)
+            coveralls=true
+            shift
+            ;;
+        --json)
+            json=true
+            shift
+            ;;
+        --pr)
+            pr=true
+            shift
+            ;;
+        *)
+            if [ -n "$pr" ]; then
+                echo "Error: --pr and package name cannot be used together"
+                exit 1
+            fi
+            searchString="$1"
+            shift
+            ;;
+    esac
+done
+
+# Get packages to test
+if [ $pr = "true" ]; then
+    base_ref=${GITHUB_BASE_REF:-main}
+    pkgs=$(git diff --name-only "origin/$base_ref"...HEAD | grep '\.go$' | xargs -n1 dirname | sort -u | uniq | xargs -I{} go list ./{} 2>/dev/null)
+elif [ -n "$searchString" ]; then
+    pkgs=$(go list ./... | grep -v /vendor/ | grep "$searchString")
+else
+    pkgs=$(go list ./... | grep -v /vendor/)
+fi
+
+if [ $json = "true" ]; then
+    echo "Generating JSON coverage data"
+    generate_cover_data_json $pkgs
+else
+    echo "Generating coverage data"
+    generate_cover_data $pkgs
+fi
+
 show_cover_report func
-case "$1" in
-"")
-    ;;
---html)
-    show_cover_report html ;;
---coveralls)
-    push_to_coveralls ;;
-*)
-    echo >&2 "error: invalid option: $1"; exit 1 ;;
-esac
+
+if [ $html = "true" ]; then
+    echo "Generating HTML report"
+    show_cover_report html
+    echo "HTML report generated at $workdir/cover.html"
+fi
+if [ $coveralls = "true" ]; then
+    push_to_coveralls
```

Comment:
@cipherboy The coveralls app would have to be added to the repo by one of the repo admins.

Reply:
@phyrog I think that should be fine; it looks like it is free and publicly visible so it isn't limited to org-connected individuals.

Let me know if there's any objections @DrDaveD @JanMa @DanGhita @naphelps

Reply:
No objections

Reply:
I think that's quorum between this and ML. @rencooo / @phyrog -- do you have docs?

Reply:
@cipherboy Just this page: https://docs.coveralls.io/index

From then on the script in the GitHub action should run goveralls to push the coverage to coveralls. Not sure if that already includes PR comments etc, but IMO we can find that out once coveralls is integrated.